feat(core): isEnabled via commands with dispatch=undefined

tomas-c opened the pull request:
## Description

This PR changes code related to Extension.isActive() and Extension.isEnabled(). 

The action.isActive() and action.isEnabled() behaviour remains the same so this should not break code that doesn't access extensions directly.

### Change 1
Before this PR extensions by default would return true when calling Extension.isEnabled() and most extensions never implemented a custom Extension.isEnabled(). This resulted in issues such as #226 where action.isEnabled() would be true but command could not be run successfully. 

Extensions no longer need to implement isEnabled. It is provided by extension manager by calling commands with dispatch=undefined [as Prosemirror intended](https://prosemirror.net/docs/guide/#commands). It's now much more difficult to have action.isEnabled() return true unless a command can be successfully run. 

### Change 2

Extension.isActive previously allowed passing in a command name argument in the same way as Extension.isEnabled. I could not find any uses of this feature. I checked scrumpy/tiptap code and they only define a single isActive method per extension, not per action command. It seems to me that the main reason why isActive allowed passing in command names was because isEnabled allowed it too.

I got rid of this feature in this PR. It doesn't change the action.isActive() signature so if people later feel this was actually useful, it can be added in back later. Unless there was a reason for this that I missed?

### ~Change 3~

~History extension allowed to override getState() and dispatch methods which made it impossible to call a command with dispatch=undefined. I could not find any uses of this feature but if this is still needed I think a better way would be to have the extension manager provide this functionality for all extensions.~
 
## Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!-- prettier-ignore-start -->

- [x] I have read the [**contributing**](https://github.com/remirror/remirror/blob/master/docs/contributing.md) document.
- [x] My code follows the code style of this project and `yarn fix` runs successfully.
- [x] I have updated the documentation where necessary.
- [x] New code is unit tested and all current tests pass when running `yarn test` .

<!-- prettier-ignore-end -->

Closes #226

tomas-c commented on @remirror/core-extensions/src/extensions/__tests__/history-extension.spec.ts:
```diff
@@ -37,61 +36,3 @@ describe('commands', () => {
-        expect(dispatcher).toHaveBeenCalledTimes(1);
-      });
-  });
-});
```
These tests were for HistoryExtension getState() and dispatch() override feature which could prevent calling commands with dispatch=undefined so I had to remove them. As I mentioned in the PR description, if this feature is needed, I suggest implementing it globally for all extensions via extension manager.

tomas-c commented on @remirror/core-extensions/src/extensions/history-extension.ts:
```diff
@@ -108,27 +61,6 @@ export class HistoryExtension extends Extension<HistoryExtensionOptions> {
-      }
-    };
-  }
-
```
This is not needed because redo() and undo() from prosemirror effectively do these checks when called with dispatch=false

tomas-c commented on @remirror/core-extensions/src/marks/link-extension.ts:
```diff
@@ -99,31 +99,26 @@ export class LinkExtension extends MarkExtension<LinkExtensionOptions> {
+      updateLink: (attrs?: Attrs): ProsemirrorCommandFunction => {
+        return (state, dispatch, view) => {
+          const { selection } = state;
+          if (selectionEmpty(selection) || (!isTextSelection(selection) && !isMarkActive({ state, type }))) {
```
You might notice that before these changes isEnabled used
```
selectionEmpty(selection) || !isTextSelection(selection)
```
but here I have
```
selectionEmpty(selection) || (!isTextSelection(selection) && !isMarkActive({ state, type }))
```

Turns out that the former would incorrectly result in isEnabled=false for this test case

tomas-c commented on @remirror/editor-wysiwyg/src/components/wysiwyg-menu.tsx:
```diff
@@ -126,7 +126,7 @@ interface MenuItemProps extends Partial<WithPaddingProps> {
   return (
     <IconButton onClick={onClick} state={state} disabled={disabled} index={index}>
-      <Icon variant={variant} />
+      <Icon variant={variant} styles={{ color: disabled ? 'gray' : undefined }} />
```
With this you can see which extensions are disabled in the wysiwyg editor

tomas-c commented:
Rebased on latest canary, now all the build checks pass

tomas-c commented:
> I'm going to need to think about the idea of removing the getState and getDispatch from the `HistoryExtension`. It's not currently being used but was added to support sub editors.

I had a few thoughts here, tell me if they make sense:

1. Is it just the history extension that might need this feature or could others too? Would it make sense to come up with a global solution?
2. Would this be solved if extension manager provided access to the command, for example `action.command(state, dispatch, view)` in addition to `action()`, `action.isEnabled()` and `action.isActive()`?
3. Would this be solved if extension manager allowed to globally overwrite dispatch and getState method so that if you do overwrite it, all extensions still get the same state and dispatch functions? This would effectively move the wrapMethod logic from history extension to extension manager.

Is there an example implementation with a sub-editor? Would be happy to take a look and see if I can come up with other options. Difficult to reason about this without trying to implement a sub-editor. If there isn't an example, we could easily add this back in when it becomes needed. Perhaps then it will be clear what's the optimal way.

Also, I could also still make this PR work while keeping the wrapMethod like so:

```
private readonly wrapMethod = (method: ProsemirrorCommandFunction): ProsemirrorCommandFunction => {	
    return (state, dispatch, view) => {	
      const { getState, getDispatch } = this.options;	

      const wrappedState = isFunction(getState) ? getState() : state;	
      const wrappedDispatch = isFunction(getDispatch) && dispatch ? getDispatch() : dispatch;	

      return method(wrappedState, wrappedDispatch, view);	
    };	
  };
```

but then it feels wrong that this is the only extension which does this wraping

ifiokjr commented:
> Is there an example implementation with a sub-editor? Would be happy to take a look and see if I can come up with other options. Difficult to reason about this without trying to implement a sub-editor. If there isn't an example, we could easily add this back in when it becomes needed. Perhaps then it will be clear what's the optimal way.

There's no internal example, but here's [an example using prosemirror](https://prosemirror.net/examples/footnote/). 

> Would this be solved if extension manager allowed to globally overwrite dispatch and getState method so that if you do overwrite it, all extensions still get the same state and dispatch functions?

It's very specific to the history extension. Since you only ever want one history stack. Other extensions don't ever really need this.

I'm happy to merge this if you include that fix you mentioned to keep the history extension support for sub editors.

tomas-c commented on docs/CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 # @remirror/docs
 
 ## 0.0.1
+
```
Running `yarn fix` added this

tomas-c commented:
> It's very specific to the history extension. Since you only ever want one history stack. Other extensions don't ever really need this.
> 
> I'm happy to merge this if you include that fix you mentioned to keep the history extension support for sub editors.

Makes sense. I have rebased on canary and reverted those getState and dispatch changes.
